File: probs/17298.py
import sys
input = sys.stdin.readline

input = open('input.txt', 'r').readline

from collections import deque

N = int(input())
A = list(map(int, input().split()))

s = deque()
ans = [0]*N

for i in reversed(range(N)):
    while s and s[-1] <= A[i]:
        s.pop()
    ans[i] = s[-1] if s else -1
    s.append(A[i])
print(*ans)

File: probs/17298.cpp
#include <bits/stdc++.h>

using namespace std;

const int MAX = 1e6;

int A[MAX], ans[MAX];

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    freopen("input.txt", "r", stdin);

    int N; cin >> N;
    for (int i = 0; i < N; i++) cin >> A[i];

    stack<int> S;
    for (int i = N-1; i >= 0; i--) {
        while (!S.empty() && S.top() <= A[i]) S.pop();
        ans[i] = S.empty() ? -1 : S.top();
        S.push(A[i]);
    }

    for (int i = 0; i < N; i++) cout << ans[i] << ' ';
    cout << '\n';

    return 0;
}
